working/4200-metaobjects/feature-specification.md

```diff
@@ -0,0 +1,519 @@
+class for each class that has a `static implements` clause, known as the
+_metaobject class_. The metaobject class implements all members of the
+specified interface by forwarding to a static member or a constructor of
+the underlying class, mixin, mixin class, or enum declaration.
+
+For example:
+
+```dart
+class MetaobjectForA implements Type, Fooable<A> {
+  const MetaobjectForA();
+  String foo(A a) => A.foo(a);
+  bool operator ==(Object other) => ...;
+  int get hashCode => ...;
+}
+
+class MetaobjectForB implements Type, Fooable<B> {
+  const MetaobjectForB();
+  String foo(B b) => B.foo(b);
+  bool operator ==(Object other) => ...;
+  int get hashCode => ...;
+}
+```
+
+These classes are implicitly generated by the compiler and analyzer, so we
+can't refer to them in user code (the real name is probably something like
+`$xBz8_`, and it's certainly a fresh name such that it doesn't coincide
+with any name that a developer has written). However, when a type variable
+like `X` above is evaluated as an expression, the resulting metaobject will
+be an instance of one of these classes.
+
+These classes implement operator `==` and the getter `hashCode`, such that
+they can behave the same as objects of type `Type` do today, when they are
+obtained by evaluating a type literal as an expression.
+
+The static type of the metaobject will be the metaobject class when the
+type literal which is being evaluated as an expression is a compile-time
+constant type (for example, `var myMetaObject = A;`). In the case where the
+type literal is a type variable `X` that has a bound of the form
+`static extends I`, the metaobject has a static type which is a subtype of
+`Type` and a subtype of `I`. This implies that we can use the members of
+`I` on that metaobject. E.g., we can call `X.foo(x)` in the example above.
+
+As a special case (ensuring that this feature does not break existing
+code), the result of evaluating a type literal that denotes a class, a
+mixin, a mixin class, or an enum declaration that does _not_ have a
+`static implements` or `static extends` clause has static type `Type`, and
+it works exactly the same as today. So does the result of evaluating a type
+that does not denote a declaration (that is, a function type, a record
+type, special types like `dynamic`, union types like `T?`  and
+`FutureOr<T>`, etc.)
+
+The previous example showed how we can use metaobjects to provide access to
+static members of a set of classes (or mixins, etc.) without depending on
+the exact class (mixin, etc.) declaration. This basically means that we
+have made the static members _virtual_, because we're calling them via
+normal instance members (that is, virtual members, with all the usual OO
+semantics) of the metaobject.
+
+The next example illustrates how we can use metaobjects to call
+constructors in a similar way (yielding 'virtual constructors'):
+
+```dart
+abstract class Creation<X> {
```

Comment:
Maybe we should be allowed to declare a `static class ...` which cannot be instantiated (it has a freshly-named zero-argument `const` constructor) and must be constant.
It can only be used for `static extends`, maybe `static implements`.